src/encoding.rs

```diff
@@ -2,69 +2,62 @@
 //! All methods err using VMError::FormatError for convenience.
 
 use byteorder::{ByteOrder, LittleEndian};
+use core::ops::Range;
 use curve25519_dalek::ristretto::CompressedRistretto;
 use curve25519_dalek::scalar::Scalar;
 
 use crate::errors::VMError;
 
-/// Reads a byte and returns the subsequent slice
-pub fn read_u8<'a>(slice: &'a [u8]) -> Result<(u8, &'a [u8]), VMError> {
-    if slice.len() < 1 {
-        return Err(VMError::FormatError);
-    }
-    Ok((slice[0], &slice[1..]))
+pub struct SliceView<'a,T> {
+    whole: &'a [T],
+    range: Range<usize>
 }
 
-/// Reads the LE32-encoded integer and returns the subsequent slice
-pub fn read_u32<'a>(slice: &'a [u8]) -> Result<(u32, &'a [u8]), VMError> {
-    if slice.len() < 4 {
-        return Err(VMError::FormatError);
+impl<'a, T> SliceView<'a,T> {
+    // get returns the slice from the given range, advancing the
+    // internal range
+    fn get(mut self, offset: usize) -> Result<&'a [T], VMError> {
```

Comment:
Shouldnt the result be Self, not native slice?

Reply:
And separate `impl From<sliceview> for &[u8]` for conversion after reslicing.

Reply:
my thought was that this is how you get the slice of data, and then we modify `Self` to update the range, but that might not map on well to how we want to access the SliceView

Reply:
Thinking about it, i suggest the following:

1. `split_at(offset)` method similar to the one in slice api. It will consume sliceview and return two sliceviews: before and after offset.
2. `Deref<&[u8]>` trait impl for sliceview for automatic conversion to a normal slice w/o explicit .into().